### ISSUE TITLE
Neaten up routing

### ISSUE THREAD

**Issue by dajinchu** [ryan-bradford/WHAM#9]
I think we could neaten the way we route things in a multitude of ways. None of this is urgent or necessary, but could definitely help us keep the code organized and easy to work with, which helps us write safe, robust code more quickly, and would be additionally valuable if we ever have more people joining this project.

## Separating JSON api from HTML rendering

We might want to move all the JSON api routes to start with /api. This should reduce clutter all around. We can also then use express.Router() to keep everything api related self contained in it's own module. Right now we've got a ton of functions in files in the /main directory that we require into index.js. Using routers would reduce all that to `app.use('/api', router)`

[More information about routers](https://expressjs.com/en/api.html#router)

I just did something like this at work, and it really neatened things up. 

Another simple benefit is we won't have to have a million if else clauses to figure out whether to render json or html.

## Taking routers a step further

Building off the idea of using express.Router() to neaten up the API, we can actually nest routers within our REST API. Every resource in our API eg. /patient, /therapist, /messages, /sessions, can have it's own router. A router can `.use` another router, allowing us to do stuff like /patient/:id/sessions by just doing something like `patientRouter.use('/:patientId/sessions', sessionRouter)` 
There is a small chance this could let us reuse routers, so /patients and /therapists/:id/patients could use the same router, though in all likelihood those two resources are too different.
Either way, it should help organize code.

[A medium length read about this technique](https://scotch.io/tutorials/keeping-api-routing-clean-using-express-routers)

## Using Express Middleware 

I admit I'm still catching up on understanding all of our code, so this idea isn't 100% fleshed out, but I think we can use express middleware as an elegant and more standard solution replacing the authorizer and responder. 

[Information about middleware](https://expressjs.com/en/guide/using-middleware.html)

[A very nice example of doing authorization with middleware](https://gist.github.com/joshnuss/37ebaf958fe65a18d4ff)

If we separate the api out, and use routers, we can give each router it's own error handlers and authorization middleware. For example for /patient, we could do have the patient router have a middleware function that runs before all of the route handlers in /patient that checks to make sure the logged in user is actually the patient in question. 
We can also, across the api, set up error handlers to send out errors the way we want, instead of using the responder class.

## Simplify HTML routes (Single Page App?)

After pulling the API out into /api we might want to move all the HTML pages to the root, ie. wham.com/patient/id, wham.com/therapist/id. Both of which are a dashboard when the user is looking at their own page, and if a therapist wants to see patient data, they can go to /patient/id to see information.

I'm starting to lean towards having a dynamic site with client side code the more I think about it. In a single page application model, we would just have /patient and /therapist and they would display a dashboard. There would be no other html routes, and the client side javascript code would just query the API for data about messages, patients, etc. In this version, we would only have to worry about the security of the JSON API, rather than the security of the JSON and the HTML, and it would give us a single point of control through the API. We basically just develop the patient dashboard and therapist dashboard as two different clients of the API.

# Conclusion
Sorry I just wrote a novel and linked an hour's worth of reading. Again, this isn't required for WHAM to work, so we can work through this slowly. Let me know what you think of these ideas, because you're definitely more familiar with the current state of the code.

**Comment by ryan-bradford** [comment]
Thanks for the good reads!

I don't totally understand routers, so correct me if I'm wrong. It feels like the system of helper functions located in the '/main/...' directory are almost equivalent to routers, just not using Express's '.use' notation. From what I understand, if we restructured our app to use Routers, we would just end up having the same amount of helper JS files (roughly) and instead of having

app.get('/patients',...) and app.post('patients',...) 

We would instead do 

app.use(routerForSlashPatient);

I feel like I'm misunderstanding routers though, especially because express makes such a big deal of them, so totally correct me if I'm wrong. 

Oh the front of middleware, this seems like a great idea. I noticed as I was coding I was copy/pasting a lot of the same stuff, and I feel like middleware can simplify this. I'm gonna work on implementing some of that right now!

Finally, for the single page application part, I'm not totally sure. The thing I like right now is that the HTML part so simply integrates with Express. We just have Handlebars render the JSON content (or some form of it). Because the HTML rendering is so tightly integrated with the JSON API, there really is no security risk. However, I haven't worked enough on the Webpage side of the app dev yet to really know what limitations that using Express presents.

**Comment by dajinchu** [comment]
These are good questions. I think you are right that what we have with the functions in /main is similar to what routers would do. From what I understand, if we start using routers, we could still keep those functions in /main, but we would just compose them together with the routers. The advantage here is that we can write router-specific middleware to reuse code across a resource. For example, we could apply the canViewPatient authorization middleware for the entire patient resource with `patientRouter.use()`

I think putting html rendering on the client shouldn't be too difficult because we can use handlebars in browsers too. Maybe I can give the single page application approach a try as a feasibility test?

**Comment by ryan-bradford** [comment]
Ahhh I think it all makes a little more sense now. I can work on transition to using routers tomorrow/Wednesday. And yea, when you get the chance, try out the single page application and see where you get. For that design, would we probably have to use some other framework like Angular or the like?

**Comment by ryan-bradford** [comment]
So I implemented the routers. It's looking good. Definitely made the Middleware more useful overall. 

I did it so we have 4 routers:

1. main - delegates all requests to the api 
     It calls either one of the three routers or a few preset URLS (/register, /logout, /api, /patients (POST), and /therapists (POST)
2. login - delegates all requests to /login
3. patient - delegates (mostly) all requests to /patients
4. therapist - delegates (mostly) all requests to /therapists

The routers rely on app.param('name', function) for authorizing which users are trying to access what IDs. 

I had to change the API just a little and move the request accept from /therapist/ID/patients/ID (PATCH) to /patients/ID/therapists/ID (PATCH) for authorization reasons. 

Every router still delegates to one of the helpers located in /main.
